src/dune_pkg/rev_store.mli

```diff
@@ -12,44 +35,84 @@ module Object : sig
 
   val of_sha1 : string -> t option
   val to_hex : t -> string
-  val equal : t -> t -> bool
-  val to_dyn : t -> Dyn.t
 end
 
 module File : sig
+  (** A file entry at a particular revision. May be a direct blob or a redirect
+      into a submodules file. *)
   type t
 
-  val path : t -> Path.Local.t
   val to_dyn : t -> Dyn.t
 
+  (** Path relative to the repository root at the given revision. *)
+  val path : t -> Path.Local.t
+
   module Set : Set.S with type elt = t
 end
 
+(** [content_of_files t files] fetches the content of a list of [files] in one
+    batch. It returns a list of strings consisting of the content of the
+    corresponding files. *)
+val content_of_files : t -> File.t list -> string list Fiber.t
+
+module Remote : sig
+  (** A git remote repository known to the revision store. *)
+  type t
+
+  val to_dyn : t -> Dyn.t
+
+  (** The [default_branch] of a remote repository. *)
+  val default_branch : t -> Object.resolved option Fiber.t
+end
+
+(** [remote ~loc ~url] gets the remote pointed to by [url] and makes
+    the revision store aware of it. This is idempotent and is memoized by
+    [url]. *)
+val remote : t -> loc:Loc.t -> url:string -> Remote.t
+
 module At_rev : sig
+  (** An immutable Git repository view at a specific commit. *)
   type t
 
-  module Config : sig
-    val parse : string -> (string * string option * string * string) option
-  end
+  val equal : t -> t -> bool
 
+  (** The underlying commit. *)
   val rev : t -> Object.t
+
+  (** Read file content at this revision. Returns [None] if the file does not
+      exist or is not a blob. *)
   val content : t -> Path.Local.t -> string option Fiber.t
+
+  (** [directory_entries t ~recursive path] lists the files under [path]. If
+      [recursive] then lists the full subtree. Symbolic links are returned as
+      files. *)
   val directory_entries : t -> recursive:bool -> Path.Local.t -> File.Set.t
-  val equal : t -> t -> bool
-  val check_out : t -> target:Path.t -> unit Fiber.t
-end
 
-module Remote : sig
-  (** handle representing a particular git repository *)
-  type t
+  (** [check_out t ~target] materialises the entire tree into [target]
+      including any submodules. *)
+  val check_out : t -> target:Path.t -> unit Fiber.t
```

Comment:
Why is this needed? I think one of the ideas of the revision store is that we never need to materialize anything.

Reply:
This is how we materialise sources in the package rules I believe. In Git we also call switching to a branch checking out, but this is really about the more specific operation of choosing a revision and a file and materialising it.

Reply:
Yes, but checking out materializes a work tree, which the rev store shouldn't do (at least not for opam repositories). But yeah, maybe it is used for package sources that are pinned via git, that would make sense.

Reply:
Just so that we are on the same page, we are talking about materialising sources somewhere else not inside the rev_store right?

Reply:
Yes. The rev store is always just a bare git repo without a worktree (which is by design, to allow concurrent access, without fighting which worktree to materialize).